proposals/p6333.md

```diff
@@ -0,0 +1,586 @@
+
+-   `carbon_library` similar to `cc_library`.
+-   `carbon_binary` similar to `cc_binary`.
+-   In the `core` directory, `carbon_library` rules that assist with builds.
+    -   The prelude would be a default dependency of all `carbon_library` rules.
+
+These are intended to accept flexible inputs:
+
+-   Support passing in standard C++ file extensions to any of these for
+    compilation.
+-   For `carbon build` in particular, it should not be necessary to pass in
+    `Core` files that are required.
+    -   We will require a correlation between library names inside `Core` and
+        directory structure. For example, `prelude/types`
+        [maps to](#mapping-packaging-directives-to-filenames)
+        `core/prelude/types.carbon`.
+    -   The same strict correlation will be supported for other packages.
+
+At the end, it should be possible to:
+
+-   Run `carbon build main.carbon` (or even just `carbon build`) with
+    non-prelude `Core` imports, and get an executable `program`.
+-   Have Bazel rules that mix C++ code and Carbon code. For example:
+
+    ```bazel
+    carbon_library(
+        name = "foo",
+        srcs = ["foo.cpp", "foo.impl.carbon"],
+        apis = ["foo.carbon"],
+    )
+    carbon_binary(
+        name = "bar",
+        srcs = ["main.cpp"],
+        deps = [":carbon_library"],
+    )
+    ```
+
+### Look-and-feel
+
+Note the goal here is to align on look-and-feel of separate compilation.
+Although the `carbon` CLI is important to the language, most details aren't
+necessary to address through the proposal process. For example, we want to get
+flag names right here, but also we wouldn't expect a proposal for flag name
+changes.
+
+## Details
+
+### Command changes
+
+#### Compile command
+
+The `carbon compile` command is intended to be a straightforward single input,
+single output command. Dependencies will be provided through a combination of:
+
+-   Given a package name to directory mapping, a
+    [filename mapping](#mapping-packaging-directives-to-filenames) based on the
+    library name.
+-   Explicitly provided input files.
+
+As part of supporting a mix of C++ and Carbon files, we will support
+`carbon compile foo.cpp` with results similar to `carbon clang -- -c foo.cpp`.
+
+#### Build command
+
+The `carbon build` command will be the new, simple way to compile, as a
+replacement for `carbon compile`. It will:
+
+-   Load provided files.
+-   For packages with directory mappings, particularly `Core`, figure out the
+    right files to load.
+-   Do something similar to the appropriate series of `carbon compile`
+    invocations.
+    -   A key divergence is that we should avoid re-checking files that would be
+        used across multiple `carbon compile` invocations.
+-   Run the equivalent of `carbon link` over produced inputs.
+
+While the build command will default to providing an executable `program`, we'll
+also want it to be capable of producing `.a` and `.so` files.
```

Comment:
@chandlerc I think you had some particular plans around `.so` files. Should we include or exclude them as possible `carbon build` outputs for now?